solver/llbsolver/ops/exec.go

```diff
@@ -90,20 +90,55 @@ func cloneExecOp(old *pb.ExecOp) pb.ExecOp {
 	n.Mounts = nil
 	for i := range old.Mounts {
 		m := *old.Mounts[i]
+
+		if m.CacheOpt != nil {
+			co := *m.CacheOpt
+			m.CacheOpt = &co
+		}
+
 		n.Mounts = append(n.Mounts, &m)
 	}
 	return n
 }
 
+func checkShouldClearCacheOpts(m *pb.Mount) bool {
+	if m.CacheOpt == nil {
+		return false
+	}
+
+	// This is a dockerfile default cache mount.
+	// We are treating this as a special case so we don't cause a cache miss unintentionally.
+	if m.CacheOpt.ID == m.Dest && m.CacheOpt.Sharing == 0 {
+		return false
+	}
+
+	// Check the case where a dockerfile cache-namespace may be used.
+	// This would be `<namespace>/<dest>`
+	_, trimmed, ok := strings.Cut(m.CacheOpt.ID, "/")
+	if ok && trimmed == m.Dest && m.CacheOpt.Sharing == 0 {
+		return false
+	}
```

Comment:
nit: can we condense these cases together? We can get the `trimmed` ID, and then compare and check the sharing, given they're essentially the same case.

I'm also still not sure I understand where this special case can actually happen. Is it just to prevent dockerfile cache misses when upgrading buildkit?

Reply:
> can we condense these cases together? We can get the trimmed ID, and then compare and check the sharing, given they're essentially the same case.

Seems like this would make the code more confusing, and certainly more expensive in the case where there is nothing to trim.

Reply:
> Is it just to prevent dockerfile cache misses when upgrading buildkit?

Yes, I believe so.
See the discussion thread here: https://dockercommunity.slack.com/archives/C7S7A40MP/p1705105919498119